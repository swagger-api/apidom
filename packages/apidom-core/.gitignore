/dist
/es
/cjs
/types
/NOTICE
/swagger-api-apidom-core-*.tgz
