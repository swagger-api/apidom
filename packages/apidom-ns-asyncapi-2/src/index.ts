import './refractor/registration';

export {
  isRefElement,
  isLinkElement,
  isMemberElement,
  isObjectElement,
  isArrayElement,
  isBooleanElement,
  isNullElement,
  isElement,
  isNumberElement,
  isStringElement,
} from '@swagger-api/apidom-core';

export { default as mediaTypes, AsyncAPIMediaTypes } from './media-types';

// eslint-disable-next-line no-restricted-exports
export { default } from './namespace';

export { default as refractorPluginReplaceEmptyElement } from './refractor/plugins/replace-empty-element';

export {
  isAsyncApi2Element,
  isAsyncApiVersionElement,
  isChannelBindingsElement,
  isChannelItemElement,
  isChannelItemElementExternal,
  isChannelsElement,
  isComponentsElement,
  isContactElement,
  isIdentifierElement,
  isInfoElement,
  isLicenseElement,
  isOperationElement,
  isParameterElement,
  isParametersElement,
  isReferenceElement,
  isReferenceElementExternal,
  isSchemaElement,
  isBooleanJsonSchemaElement,
  isSecurityRequirementElement,
  isServerElement,
  isServerBindingsElement,
  isServersElement,
  isServerVariableElement,
} from './predicates';

export {
  isAsyncApi2LikeElement,
  isParameterLikeElement,
  isReferenceLikeElement,
  isSchemaLikeElement,
  isSecurityRequirementLikeElement,
  isServerLikeElement,
  isAsyncApiExtension,
} from './refractor/predicates';

export { keyMap, getNodeType } from './traversal/visitor';

/**
 * AsyncApi >= 2.0.0 <=2.6.0 specification elements.
 */
export { default as AsyncApi2Element } from './elements/AsyncApi2';
export { default as AsyncApiVersionElement } from './elements/AsyncApiVersion';
export { default as ChannelBindingsElement } from './elements/ChannelBindings';
export { default as ChannelItemElement } from './elements/ChannelItem';
export { default as ChannelsElement } from './elements/Channels';
export { default as ComponentsElement } from './elements/Components';
export { default as ContactElement } from './elements/Contact';
export { default as CorrelationIDElement } from './elements/CorrelationID';
export { default as DefaultContentTypeElement } from './elements/DefaultContentType';
export { default as ExternalDocumentationElement } from './elements/ExternalDocumentation';
export { default as IdentifierElement } from './elements/Identifier';
export { default as InfoElement } from './elements/Info';
export { default as LicenseElement } from './elements/License';
export { default as MessageElement } from './elements/Message';
export { default as MessageBindingsElement } from './elements/MessageBindings';
export { default as MessageExampleElement } from './elements/MessageExample';
export { default as MessageTraitElement } from './elements/MessageTrait';
export { default as OAuthFlowElement } from './elements/OAuthFlow';
export { default as OAuthFlowsElement } from './elements/OAuthFlows';
export { default as OperationElement } from './elements/Operation';
export { default as OperationBindingsElement } from './elements/OperationBindings';
export { default as OperationTraitElement } from './elements/OperationTrait';
export { default as ParameterElement } from './elements/Parameter';
export { default as ParametersElement } from './elements/Parameters';
export { default as ReferenceElement } from './elements/Reference';
export { default as SchemaElement } from './elements/Schema';
export { default as SecurityRequirementElement } from './elements/SecurityRequirement';
export { default as SecuritySchemeElement } from './elements/SecurityScheme';
export { default as ServerElement } from './elements/Server';
export { default as ServerBindingsElement } from './elements/ServerBindings';
export { default as ServersElement } from './elements/Servers';
export { default as ServerVariableElement } from './elements/ServerVariable';
export { default as TagElement } from './elements/Tag';
export { default as TagsElement } from './elements/Tags';

/**
 * Binding elements.
 */
// AMQP 0-9-1
export { default as AmqpChannelBindingElement } from './elements/bindings/amqp/AmqpChannelBinding';
export { default as AmqpMessageBindingElement } from './elements/bindings/amqp/AmqpMessageBinding';
export { default as AmqpOperationBindingElement } from './elements/bindings/amqp/AmqpOperationBinding';
export { default as AmqpServerBindingElement } from './elements/bindings/amqp/AmqpServerBinding';
// AMQP 1.0
export { default as Amqp1ChannelBindingElement } from './elements/bindings/amqp1/Amqp1ChannelBinding';
export { default as Amqp1MessageBindingElement } from './elements/bindings/amqp1/Amqp1MessageBinding';
export { default as Amqp1OperationBindingElement } from './elements/bindings/amqp1/Amqp1OperationBinding';
export { default as Amqp1ServerBindingElement } from './elements/bindings/amqp1/Amqp1ServerBinding';
// Anypoint MQ
export { default as AnypointmqChannelBindingElement } from './elements/bindings/anypointmq/AnypointmqChannelBinding';
export { default as AnypointmqMessageBindingElement } from './elements/bindings/anypointmq/AnypointmqMessageBinding';
export { default as AnypointmqOperationBindingElement } from './elements/bindings/anypointmq/AnypointmqOperationBinding';
export { default as AnypointmqServerBindingElement } from './elements/bindings/anypointmq/AnypointmqServerBinding';
// Google Cloud Pub/Sub
export { default as GooglepubsubChannelBindingElement } from './elements/bindings/googlepubsub/GooglepubsubChannelBinding';
export { default as GooglepubsubMessageBindingElement } from './elements/bindings/googlepubsub/GooglepubsubMessageBinding';
export { default as GooglepubsubOperationBindingElement } from './elements/bindings/googlepubsub/GooglepubsubOperationBinding';
export { default as GooglepubsubServerBindingElement } from './elements/bindings/googlepubsub/GooglepubsubServerBinding';

// HTTP
export { default as HttpChannelBindingElement } from './elements/bindings/http/HttpChannelBinding';
export { default as HttpMessageBindingElement } from './elements/bindings/http/HttpMessageBinding';
export { default as HttpOperationBindingElement } from './elements/bindings/http/HttpOperationBinding';
export { default as HttpServerBindingElement } from './elements/bindings/http/HttpServerBinding';
// IBM MQ
export { default as IbmmqChannelBindingElement } from './elements/bindings/ibmmq/IbmmqChannelBinding';
export { default as IbmmqMessageBindingElement } from './elements/bindings/ibmmq/IbmmqMessageBinding';
export { default as IbmmqOperationBindingElement } from './elements/bindings/ibmmq/IbmmqOperationBinding';
export { default as IbmmqServerBindingElement } from './elements/bindings/ibmmq/IbmmqServerBinding';
// JMS
export { default as JmsChannelBindingElement } from './elements/bindings/jms/JmsChannelBinding';
export { default as JmsMessageBindingElement } from './elements/bindings/jms/JmsMessageBinding';
export { default as JmsOperationBindingElement } from './elements/bindings/jms/JmsOperationBinding';
export { default as JmsServerBindingElement } from './elements/bindings/jms/JmsServerBinding';
// Kafka
export { default as KafkaChannelBindingElement } from './elements/bindings/kafka/KafkaChannelBinding';
export { default as KafkaMessageBindingElement } from './elements/bindings/kafka/KafkaMessageBinding';
export { default as KafkaOperationBindingElement } from './elements/bindings/kafka/KafkaOperationBinding';
export { default as KafkaServerBindingElement } from './elements/bindings/kafka/KafkaServerBinding';
// Mercure
export { default as MercureChannelBindingElement } from './elements/bindings/mercure/MercureChannelBinding';
export { default as MercureMessageBindingElement } from './elements/bindings/mercure/MercureMessageBinding';
export { default as MercureOperationBindingElement } from './elements/bindings/mercure/MercureOperationBinding';
export { default as MercureServerBindingElement } from './elements/bindings/mercure/MercureServerBinding';
// MQTT
export { default as MqttChannelBindingElement } from './elements/bindings/mqtt/MqttChannelBinding';
export { default as MqttMessageBindingElement } from './elements/bindings/mqtt/MqttMessageBinding';
export { default as MqttOperationBindingElement } from './elements/bindings/mqtt/MqttOperationBinding';
export { default as MqttServerBindingElement } from './elements/bindings/mqtt/MqttServerBinding';
// MQTT 5
export { default as Mqtt5ChannelBindingElement } from './elements/bindings/mqtt5/Mqtt5ChannelBinding';
export { default as Mqtt5MessageBindingElement } from './elements/bindings/mqtt5/Mqtt5MessageBinding';
export { default as Mqtt5OperationBindingElement } from './elements/bindings/mqtt5/Mqtt5OperationBinding';
export { default as Mqtt5ServerBindingElement } from './elements/bindings/mqtt5/Mqtt5ServerBinding';
// NATS
export { default as NatsChannelBindingElement } from './elements/bindings/nats/NatsChannelBinding';
export { default as NatsMessageBindingElement } from './elements/bindings/nats/NatsMessageBinding';
export { default as NatsOperationBindingElement } from './elements/bindings/nats/NatsOperationBinding';
export { default as NatsServerBindingElement } from './elements/bindings/nats/NatsServerBinding';
// Pulsar
export { default as PulsarChannelBindingElement } from './elements/bindings/pulsar/PulsarChannelBinding';
export { default as PulsarMessageBindingElement } from './elements/bindings/pulsar/PulsarMessageBinding';
export { default as PulsarOperationBindingElement } from './elements/bindings/pulsar/PulsarOperationBinding';
export { default as PulsarServerBindingElement } from './elements/bindings/pulsar/PulsarServerBinding';
// Redis
export { default as RedisChannelBindingElement } from './elements/bindings/redis/RedisChannelBinding';
export { default as RedisMessageBindingElement } from './elements/bindings/redis/RedisMessageBinding';
export { default as RedisOperationBindingElement } from './elements/bindings/redis/RedisOperationBinding';
export { default as RedisServerBindingElement } from './elements/bindings/redis/RedisServerBinding';
// SNS
export { default as SnsChannelBindingElement } from './elements/bindings/sns/SnsChannelBinding';
export { default as SnsMessageBindingElement } from './elements/bindings/sns/SnsMessageBinding';
export { default as SnsOperationBindingElement } from './elements/bindings/sns/SnsOperationBinding';
export { default as SnsServerBindingElement } from './elements/bindings/sns/SnsServerBinding';
// Solace
export { default as SolaceChannelBindingElement } from './elements/bindings/solace/SolaceChannelBinding';
export { default as SolaceMessageBindingElement } from './elements/bindings/solace/SolaceMessageBinding';
export { default as SolaceOperationBindingElement } from './elements/bindings/solace/SolaceOperationBinding';
export { default as SolaceServerBindingElement } from './elements/bindings/solace/SolaceServerBinding';
// SQS
export { default as SqsChannelBindingElement } from './elements/bindings/sqs/SqsChannelBinding';
export { default as SqsMessageBindingElement } from './elements/bindings/sqs/SqsMessageBinding';
export { default as SqsOperationBindingElement } from './elements/bindings/sqs/SqsOperationBinding';
export { default as SqsServerBindingElement } from './elements/bindings/sqs/SqsServerBinding';
// STOMP
export { default as StompChannelBindingElement } from './elements/bindings/stomp/StompChannelBinding';
export { default as StompMessageBindingElement } from './elements/bindings/stomp/StompMessageBinding';
export { default as StompOperationBindingElement } from './elements/bindings/stomp/StompOperationBinding';
export { default as StompServerBindingElement } from './elements/bindings/stomp/StompServerBinding';
// WebSocket
export { default as WebSocketChannelBindingElement } from './elements/bindings/ws/WebSocketChannelBinding';
export { default as WebSocketMessageBindingElement } from './elements/bindings/ws/WebSocketMessageBinding';
export { default as WebSocketOperationBindingElement } from './elements/bindings/ws/WebSocketOperationBinding';
export { default as WebSocketServerBindingElement } from './elements/bindings/ws/WebSocketServerBinding';
