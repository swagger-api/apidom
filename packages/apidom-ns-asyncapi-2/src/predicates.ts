import {
  createPredicate,
  isBooleanElement,
  isStringElement,
  toValue,
} from '@swagger-api/apidom-core';

import AsyncApi2Element from './elements/AsyncApi2';
import AsyncApiVersionElement from './elements/AsyncApiVersion';
import ChannelBindingsElement from './elements/ChannelBindings';
import ChannelItemElement from './elements/ChannelItem';
import ChannelsElement from './elements/Channels';
import ComponentsElement from './elements/Components';
import ContactElement from './elements/Contact';
import IdentifierElement from './elements/Identifier';
import InfoElement from './elements/Info';
import LicenseElement from './elements/License';
import OperationElement from './elements/Operation';
import ParameterElement from './elements/Parameter';
import ParametersElement from './elements/Parameters';
import ReferenceElement from './elements/Reference';
import SchemaElement from './elements/Schema';
import SecurityRequirementElement from './elements/SecurityRequirement';
import ServerElement from './elements/Server';
import ServerBindingsElement from './elements/ServerBindings';
import ServersElement from './elements/Servers';
import ServerVariableElement from './elements/ServerVariable';

export const isAsyncApi2Element = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq, hasClass }) => {
    return (element: any) =>
      element instanceof AsyncApi2Element ||
      (hasBasicElementProps(element) &&
        isElementType('asyncApi2', element) &&
        primitiveEq('object', element) &&
        hasClass('api', element));
  },
);

export const isAsyncApiVersionElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof AsyncApiVersionElement ||
      (hasBasicElementProps(element) &&
        isElementType('asyncApiVersion', element) &&
        primitiveEq('string', element));
  },
);

export const isChannelBindingsElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ChannelBindingsElement ||
      (hasBasicElementProps(element) &&
        isElementType('channelBindings', element) &&
        primitiveEq('object', element));
  },
);

export const isChannelItemElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ChannelItemElement ||
      (hasBasicElementProps(element) &&
        isElementType('channelItem', element) &&
        primitiveEq('object', element));
  },
);

export const isChannelItemElementExternal = (element: any): element is ChannelItemElement => {
  if (!isChannelItemElement(element)) {
    return false;
  }
  if (!isStringElement(element.$ref)) {
    return false;
  }

  const value = toValue(element.$ref);

  return typeof value === 'string' && value.length > 0 && !value.startsWith('#');
};

export const isChannelsElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ChannelsElement ||
      (hasBasicElementProps(element) &&
        isElementType('channels', element) &&
        primitiveEq('object', element));
  },
);

export const isComponentsElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ComponentsElement ||
      (hasBasicElementProps(element) &&
        isElementType('components', element) &&
        primitiveEq('object', element));
  },
);

export const isContactElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ContactElement ||
      (hasBasicElementProps(element) &&
        isElementType('contact', element) &&
        primitiveEq('object', element));
  },
);

export const isIdentifierElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof IdentifierElement ||
      (hasBasicElementProps(element) &&
        isElementType('identifier', element) &&
        primitiveEq('string', element));
  },
);

export const isInfoElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof InfoElement ||
      (hasBasicElementProps(element) &&
        isElementType('info', element) &&
        primitiveEq('object', element));
  },
);

export const isLicenseElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof LicenseElement ||
      (hasBasicElementProps(element) &&
        isElementType('license', element) &&
        primitiveEq('object', element));
  },
);

export const isOperationElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof OperationElement ||
      (hasBasicElementProps(element) &&
        isElementType('operation', element) &&
        primitiveEq('object', element));
  },
);

export const isParameterElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ParameterElement ||
      (hasBasicElementProps(element) &&
        isElementType('parameter', element) &&
        primitiveEq('object', element));
  },
);

export const isParametersElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ParametersElement ||
      (hasBasicElementProps(element) &&
        isElementType('parameters', element) &&
        primitiveEq('object', element));
  },
);

export const isReferenceElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ReferenceElement ||
      (hasBasicElementProps(element) &&
        isElementType('reference', element) &&
        primitiveEq('object', element));
  },
);

export const isReferenceElementExternal = (element: any): element is ReferenceElement => {
  if (!isReferenceElement(element)) {
    return false;
  }

  const value = toValue(element.$ref);

  return typeof value === 'string' && value.length > 0 && !value.startsWith('#');
};

export const isSchemaElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof SchemaElement ||
      (hasBasicElementProps(element) &&
        isElementType('schema', element) &&
        primitiveEq('object', element));
  },
);

export const isBooleanJsonSchemaElement = (element: any) => {
  return isBooleanElement(element) && element.classes.includes('boolean-json-schema');
};

export const isSecurityRequirementElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof SecurityRequirementElement ||
      (hasBasicElementProps(element) &&
        isElementType('securityRequirement', element) &&
        primitiveEq('object', element));
  },
);

export const isServerElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ServerElement ||
      (hasBasicElementProps(element) &&
        isElementType('server', element) &&
        primitiveEq('object', element));
  },
);

export const isServerBindingsElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ServerBindingsElement ||
      (hasBasicElementProps(element) &&
        isElementType('serverBindings', element) &&
        primitiveEq('object', element));
  },
);

export const isServersElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ServersElement ||
      (hasBasicElementProps(element) &&
        isElementType('servers', element) &&
        primitiveEq('object', element));
  },
);

export const isServerVariableElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ServerVariableElement ||
      (hasBasicElementProps(element) &&
        isElementType('serverVariable', element) &&
        primitiveEq('object', element));
  },
);
