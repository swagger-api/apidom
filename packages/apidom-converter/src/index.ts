import { ParseResultElement } from '@swagger-api/apidom-core';
import { mergeOptions, bundle, File } from '@swagger-api/apidom-reference';

import defaultOptions, { ConverterOptions } from './options';
import ConvertError from './errors/ConvertError';
import UnmatchedConvertStrategyError from './errors/UnmatchedConvertStrategyError';

export { ConvertError, UnmatchedConvertStrategyError };

/**
 * `convertApiDOM` already assumes that the ApiDOM is bundled.
 */
export const convertApiDOM = async (element: ParseResultElement, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const file = File({
    uri: mergedOptions.resolve.baseURI,
    parseResult: element,
    mediaType: mergedOptions.convert.sourceMediaType || mergedOptions.parse.mediaType,
  });
  const strategy = mergedOptions.convert.strategies.find((s) => s.canConvert(file, mergedOptions));

  if (typeof strategy === 'undefined') {
    throw new UnmatchedConvertStrategyError(file.uri);
  }

  return strategy.convert(file, mergedOptions);
};

const convert = async (uri: string, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const parseResult = await bundle(uri, mergedOptions);

  return convertApiDOM(parseResult, mergedOptions);
};

export default convert;
