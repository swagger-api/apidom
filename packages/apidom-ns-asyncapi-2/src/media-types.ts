import { last } from 'ramda';
import { MediaTypes } from '@swagger-api/apidom-core';

type Format = 'generic' | 'json' | 'yaml';

export class AsyncAPIMediaTypes extends MediaTypes<string> {
  filterByFormat(format: Format = 'generic') {
    const effectiveFormat = format === 'generic' ? 'asyncapi;version' : format;
    return this.filter((mediaType) => mediaType.includes(effectiveFormat));
  }

  findBy(version = '2.4.0', format: Format = 'generic') {
    const search =
      format === 'generic'
        ? `vnd.aai.asyncapi;version=${version}`
        : `vnd.aai.asyncapi+${format};version=${version}`;
    const found = this.find((mediaType) => mediaType.includes(search));

    return found || this.unknownMediaType;
  }

  latest(format: Format = 'generic') {
    return last(this.filterByFormat(format)) as string;
  }
}

const mediaTypes = new AsyncAPIMediaTypes(
  'application/vnd.aai.asyncapi;version=2.0.0',
  'application/vnd.aai.asyncapi+json;version=2.0.0',
  'application/vnd.aai.asyncapi+yaml;version=2.0.0',
  'application/vnd.aai.asyncapi;version=2.1.0',
  'application/vnd.aai.asyncapi+json;version=2.1.0',
  'application/vnd.aai.asyncapi+yaml;version=2.1.0',
  'application/vnd.aai.asyncapi;version=2.2.0',
  'application/vnd.aai.asyncapi+json;version=2.2.0',
  'application/vnd.aai.asyncapi+yaml;version=2.2.0',
  'application/vnd.aai.asyncapi;version=2.3.0',
  'application/vnd.aai.asyncapi+json;version=2.3.0',
  'application/vnd.aai.asyncapi+yaml;version=2.3.0',
  'application/vnd.aai.asyncapi;version=2.4.0',
  'application/vnd.aai.asyncapi+json;version=2.4.0',
  'application/vnd.aai.asyncapi+yaml;version=2.4.0',
);

export default mediaTypes;
