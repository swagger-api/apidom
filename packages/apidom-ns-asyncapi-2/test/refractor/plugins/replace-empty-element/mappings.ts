import { expect } from 'chai';
import dedent from 'dedent';
import { sexprs, SourceMapElement } from '@swagger-api/apidom-core';
import { parse } from '@swagger-api/apidom-parser-adapter-yaml-1-2';

import { refractorPluginReplaceEmptyElement, AsyncApi2Element } from '../../../../src';

describe('given empty value instead of InfoElement', function () {
  it('should replace empty value with semantic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          info:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    });

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given empty value instead of ContactElement', function () {
  it('should replace empty value with semantic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          info:
            contact:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    });

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given empty value instead of Message.payload with unspecified schema format', function () {
  it('should replace empty value with semantic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          components:
            messages:
                userSignUp:
                  payload:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    });

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given empty value instead of Message.payload with supported schema format', function () {
  it('should replace empty value with semantic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          components:
            messages:
                userSignUp:
                  schemaFormat: application/vnd.aai.asyncapi;version=2.4.0
                  payload:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    });

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given empty value instead of Message.payload with unsupported schema format', function () {
  it('should replace empty value with generic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          components:
            messages:
                userSignUp:
                  schemaFormat: application/vnd.apache.avro;version=1.9.0
                  payload:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    });

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given empty value instead for AsyncAPI.components.schemas', function () {
  it('should replace empty value with semantic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          components:
            schemas:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    }) as AsyncApi2Element;
    const isComponentsSchemas = asyncApiElement
      .get('components')
      .get('schemas')
      .classes.includes('components-schemas');

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
    expect(isComponentsSchemas).to.be.true;
  });
});

describe('given empty value instead for AsyncAPI.components.schemas.*', function () {
  it('should replace empty value with semantic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          components:
            schemas:
              Schema1:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    }) as AsyncApi2Element;

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given empty value instead for Schema.properties.*', function () {
  it('should replace empty value with semantic element', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          components:
            schemas:
              User:
                properties:
                  firstName:
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    }) as AsyncApi2Element;

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given AsyncAPI definition with no empty values', function () {
  it('should do nothing', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          id: urn:com:smartylighting:streetlights:server
        `;
    const apiDOM = await parse(yamlDefinition);
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    }) as AsyncApi2Element;

    expect(sexprs(asyncApiElement)).toMatchSnapshot();
  });
});

describe('given AsyncAPI definition with empty values', function () {
  it('should generate proper source maps', async function () {
    const yamlDefinition = dedent`
          asyncapi: 2.4.0
          info:
        `;
    const apiDOM = await parse(yamlDefinition, { sourceMap: true });
    const asyncApiElement = AsyncApi2Element.refract(apiDOM.result, {
      plugins: [refractorPluginReplaceEmptyElement()],
    }) as AsyncApi2Element;
    const { info: infoValue } = asyncApiElement;
    const sourceMap = infoValue?.meta.get('sourceMap');
    const { positionStart, positionEnd } = sourceMap;
    const expectedPosition = [1, 5, 21];

    expect(infoValue?.meta.get('sourceMap')).to.be.an.instanceof(SourceMapElement);
    expect(positionStart.equals(expectedPosition)).to.be.true;
    expect(positionEnd.equals(expectedPosition)).to.be.true;
  });
});
