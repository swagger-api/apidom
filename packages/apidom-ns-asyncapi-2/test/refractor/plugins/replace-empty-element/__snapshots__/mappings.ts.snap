// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given AsyncAPI definition with no empty values should do nothing 1`] = `
(AsyncApi2Element
  (MemberElement
    (StringElement)
    (AsyncApiVersionElement))
  (MemberElement
    (StringElement)
    (IdentifierElement)))
`;

exports[`given empty value instead for AsyncAPI.components.schemas should replace empty value with semantic element 1`] = `
(AsyncApi2Element
  (MemberElement
    (StringElement)
    (AsyncApiVersionElement))
  (MemberElement
    (StringElement)
    (ComponentsElement
      (MemberElement
        (StringElement)
        (ObjectElement)))))
`;

exports[`given empty value instead for AsyncAPI.components.schemas.* should replace empty value with semantic element 1`] = `
(AsyncApi2Element
  (MemberElement
    (StringElement)
    (AsyncApiVersionElement))
  (MemberElement
    (StringElement)
    (ComponentsElement
      (MemberElement
        (StringElement)
        (ObjectElement
          (MemberElement
            (StringElement)
            (SchemaElement)))))))
`;

exports[`given empty value instead of ContactElement should replace empty value with semantic element 1`] = `
(AsyncApi2Element
  (MemberElement
    (StringElement)
    (AsyncApiVersionElement))
  (MemberElement
    (StringElement)
    (InfoElement
      (MemberElement
        (StringElement)
        (ContactElement)))))
`;

exports[`given empty value instead of InfoElement should replace empty value with semantic element 1`] = `
(AsyncApi2Element
  (MemberElement
    (StringElement)
    (AsyncApiVersionElement))
  (MemberElement
    (StringElement)
    (InfoElement)))
`;
